### DISCUSSION TITLE
Rethinking the ...Result(Item) and ...Action architecture

### DISCUSSION THREAD

**Discussion by friday** [Ulauncher/Ulauncher#966] [Ideas]
The `...Result(Item)` and `...Action` architecture is used both by extensions and Ulauncher's "modes".

Every `Result` has one or two actions (enter and alt-enter). If more than two are needed, then the second has to be to list more actions. At that point the secondary action will need its own results (and they will have actions). This leads to repetitive and confusing code.

It also leads to inconsistencies. Sometimes the list includes all the actions in the list (including the main action), and others times not. And it's hard or impossible to add dynamic behavior like implementing desktop entry actions for applications.

In addition to that, this architecture is currently implemented by extensions defining/importing the "actions", but their methods never run inside the extension. They are serialized with the rest of the object and sent to the app to run there. I haven't run any performance tests on this, but it's definitely inefficient and also a big factor that makes third party extensions for other languages impossible.

Also, since `on_enter` and `on_alt_enter` only accepts an "action", then if you want to perform two things when you select one option, you need to use an ActionList to wrap them into one action. And if you want to do a custom thing you have to make an own "action" to return that behaves very different from the other actions, since the pre-made actions are run in Ulauncher (server side) while custom actions are run in the extension. This could all be replaced by always running the callbacks on the extension side, and giving access to helper methods like "api.lib.save_to_clipboard()" or something.

**Instead...**

I want the Results to be more consistent, smarter, composable and extensible like this:
* For example a `Text` is something that is a simple node of text that has the `Copy to clipboard` action.
* A `File` is basically a path string, which means it's also a `Text`, so it can inherit from `Text` to get the `Copy to clipboard` action, but the main action would be "Open", and the second actions would be "Open with ..." showing a list of apps.
* An extension could implement a `Paste` action for the `Text` and then `File` will get it also (Paste is impossible to implement on Wayland or for people with non-standard keyboard shortcuts, so it's not something we would want in the main Ulauncher code).
* Users should be able to reorder the actions, so for example `Paste` is first for them, which also makes it the actions that triggers when they press enter.
* A Calculator result would just be a `Text` with a different icon. So if the user decides that `Paste` should be the main action, then that applies to the calculator result too.
* Other types of results that don't inherit from `Text`, like `Application` would hence not inherit the copy or paste actions either.

The modes and extensions import Result classes and returns lists of Results, which are basically just the data without any methods. This makes it lightweight to import and use these classes in extensions, and we could serialize the data to json instead of pickle, which should be easier to send to the Ulauncher extension server, and easier to implement for other programming languages if we want that.

You register action(s) for the results separately. I think a reasonable limitation is that these have static names, which means we wouldn't have to run any callback method to list the actions. Only to perform them. So for example like this:

```py
import os
import subprocess
from ulauncher.api.results import FileResult
from ulauncher.api.actions import Action, attach_action

def handler(result, action_id):
  if action_id == 'open_containing_folder':
    subprocess.Popen(['xdg-open', os.path.dirname(result.text)])  # result.text because File inherits from Text

action = Action(id='open_containing_folder', name='Open Containing Folder', icon='system-file-manager')

attach_action(FileResult, action, handler)
```

This should solve a lot of the issues, but creates new issues that need to be handled. 
* How to define the "main" and "secondary" action, especially considering you want to be able to inherit actions and make it non-default: I think this can be solved with a specific ordering where inherited actions are down-prioritized and the actions are otherwise listed in the order they are defined.
* How users override the action order and deactivate/activate actions added by extensions in particular (this needs an UI)
* Should it be possible to over-define actions by using the same names, and if not should we allow/show multiple actions with the same names?
* Extensions have to still be able to create custom `Results` with custom actions that are not accessible to Ulauncher or other extensions. So I haven't fully considered the extension integration part yet.

Dynamic actions (for example for #476) can't be handled with the same exact API. They need to have knowledge about the Result instance to generate. They can never at least safely be the "default", since different apps have different desktop entry actions. The amount of them will vary and it depends on the system and changes the app packages and users themselves can do.
I think this should not be available to extensions. Only internal use.

Example 1:

```py
import os
import subprocess
from ulauncher.api.results import FileResult
from ulauncher.api.actions import Action, attach_action_generator

def handler(result, action_id):
  path = result.text if action_id == 'open_folder' else os.path.dirname(result.text)
  subprocess.Popen(['xdg-open', path])  # Simplified for demonstrating the API

def action_generator(result):
  if os.path.isdir(result.text):
    yield Action(
      id='open_folder',
      name=f'Open Folder "{os.path.basename(result.text)}"',
      icon='system-file-manager'
    )
  if os.path.isfile(result.text):
    yield Action(
      id='open_containing_folder',
      name='Open Containing Folder',
      icon='system-file-manager'
    )

attach_action_generator(FileResult, action_generator, handler)
```

Example 2:

```py
import gi
gi.require_version('Gio', '2.0')
from gi.repository import Gio
from ulauncher.api.results import AppResult
from ulauncher.api.actions import Action, attach_action_generator

def handler(result, action_id):
  app = Gio.DesktopAppInfo.new(result.app_id)
  app.launch_action(action_id)  # Simplified for demonstrating the API

def action_generator(result):
  app = Gio.DesktopAppInfo.new(result.app_id)
  for action_id in app.list_actions():
    yield Action(
      id=action_id,
      name=app.get_action_name(action_id),
      icon=result.icon
    )

attach_action_generator(AppResult, action_generator, handler)
```

And of course the handler can return a list or results, which restarts the flow. For example:
* User enters path
* Users toggles the actions list
* User picks the action "Open With..."
* "Open With..." shows a list of apps matching the file's mimetype, that functions similar to the normal app search, but more limited and when you pick one it launches with a file.

(or the reverse flow, picking the app first and the file later)

*Note that this is a draft proposal and the code examples are just examples and doesn't consider backwards compatibility or migration path*

**Comment by brpaz** [comment]
I really like the approach of [Raycast](https://www.raycast.com/), of having a dropdown menu with the list of available actions.

![rayvast example](https://d47mokwr4vy2h.cloudfront.net/6fuoq0j51e6eu9y1kyzs3791m4lg)

Ideally, I would love Ulauncher to take some inspiration from Raycast UX. The detail views and forms support, for example, it´s something I would like to have in Ulauncher and would allow much more advanced workflows.

**Comment by friday** [comment]
> Ideally, I would love Ulauncher to take some inspiration from Raycast UX. The detail views and forms support, for example, it´s something I would like to have in Ulauncher and would allow much more advanced workflows.

The system I envisioned and tried to described would be able to handle the same workflow. But how it looks is a completely different matter and not something I would get hung up on. They're a team of 16 paid employees.

I imagined it as reusing the same window/list as for the main query, but you go "forward" one step. Then you can keep going forward and sort of "piping" the result from one extension to another. For example a Search extension would result in a file, and other extensions can add actions to files, like "Attach in Email". And this might be a multiple step process using your address book etc.

I imagine something like the file browser mode now where you can easily enter a folder or go up/back one step by pressing backspace. Would be easier to "chain".
